### PR TITLE
Replace rapidjson with nlohmann/json

### PR DESCRIPTION
**Pull request by asraa**
Commit Message: Replace rapidjson with nlohmann/json
Additional Description:
- Schema validation is not supported by the core nlohmann/json library. They endorse an option here https://github.com/nlohmann/json/issues/305 but schema validation is not used anywhere in Envoy core.
- Added runtime option `envoy.reloadable_features.remove_legacy_json`. It is disabled by default. I will enable the new parser in a separate PR.
- There is one usage in core to parse metadata in the HeaderFormatter: https://github.com/envoyproxy/envoy/blob/e073b325890542a3ce3bbc4eefb89a09f390f27e/source/common/router/header_formatter.cc#L61 This could be replaced pretty easily with protobuf's parser in a later PR
- Originally I had an implementation that shared the SAX parser between the two implementations, but RapidJson and nlohmann/json have to handle tracking line numbers differently so I split them. (RapidJSON makes it easy to subclass the stream implementation, but nlohmann/json doesn't.)
- The  major difference between the two `json_internal.cc` implementations is the custom stream class (see json_loader.cc line 243-280. It has a reference to the handler to update line numbers)

Risk Level: Low for core, maybe medium for those using JSON on the dataplane.
Release Notes: Added release notes about runtime feature
Testing:
* `json_loader_test.cc` is parametrized over both runtime values.
Fixes https://github.com/envoyproxy/envoy/issues/4705

### CODE REVIEW

**Comment by alyssawilk** [comment]
@asraa I'll let you tag a reviewer once you've got CI sorted, SG?

**Comment by asraa** [comment]
Yep! This is still in DRAFT mode, I'll move it out of draft mode when I finish some test TODOs (benchmarking)

File: bazel/external/json.BUILD [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+    hdrs = glob([
+        "include/nlohmann/*.hpp",
+        "include/nlohmann/**/*.hpp",
+        "include/nlohmann/*/*/*.hpp",
+    ]),
+    copts = [
+        "-I external/nlohmann_json_lib",
+    ],
+    visibility = ["//visibility:public"],
+    alwayslink = 1,
```
Comment (lizan) [inline bazel/external/json.BUILD:16]: no need?

File: bazel/external/json.BUILD [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+
+cc_library(
+    name = "nlohmann_json_lib",
+    hdrs = glob([
+        "include/nlohmann/*.hpp",
+        "include/nlohmann/**/*.hpp",
+        "include/nlohmann/*/*/*.hpp",
+    ]),
+    copts = [
+        "-I external/nlohmann_json_lib",
```
Comment (lizan) [inline bazel/external/json.BUILD:13]: should use `include`.

File: bazel/repository_locations.bzl [hunk 3]
```diff
@@ -445,6 +445,20 @@ REPOSITORY_LOCATIONS_SPEC = dict(
...
+        version = "3.9.1",
+        sha256 = "4cf0df69731494668bdd6460ed8cb269b68de9c19ad8c27abc24cd72605b2d5b",
+        strip_prefix = "json-{version}",
+        urls = ["https://github.com/nlohmann/json/archive/v{version}.tar.gz"],
+        # This will be a replacement for rapidJSON used in extensions and may also be a fast
+        # replacement for protobuf JSON.
+        use_category = ["controlplane", "dataplane_core"],
+        release_date = "2020-07-06",
+        cpe = "cpe:2.3:a:json_project:json:*",
+    ),
```
Comment (moderation) [inline bazel/repository_locations.bzl:461]: Are you planning on deleting the rapidjson dep?
https://github.com/envoyproxy/envoy/blob/master/bazel/repository_locations.bzl#L434-L447

**Review by moderation — CHANGES_REQUESTED** [review]

File: bazel/repository_locations.bzl (hunk 3 above)
Comment (asraa) [reply bazel/repository_locations.bzl:461]: I am planning on keeping it as a runtime switchable options in extensions -- although I am down to remove :)

File: bazel/repository_locations.bzl [hunk 4]
```diff
@@ -445,6 +445,20 @@ REPOSITORY_LOCATIONS_SPEC = dict(
...
+    com_github_nlohmann_json = dict(
+        project_name = "nlohmann JSON",
+        project_desc = "Fast JSON parser/generator for C++",
+        project_url = "https://nlohmann.github.io/json",
+        version = "3.9.1",
+        sha256 = "4cf0df69731494668bdd6460ed8cb269b68de9c19ad8c27abc24cd72605b2d5b",
+        strip_prefix = "json-{version}",
+        urls = ["https://github.com/nlohmann/json/archive/v{version}.tar.gz"],
+        # This will be a replacement for rapidJSON used in extensions and may also be a fast
+        # replacement for protobuf JSON.
```
Comment (dio) [inline bazel/repository_locations.bzl:457]: Cool!

**Comment by asraa** [comment]
There is an existing clang-tidy issue about a potential memory leak in the constructor of `rapidjson::SchemaDoument` https://github.com/envoyproxy/envoy/blob/12c42b599b053707dcfd13dfdd87c3cd75c95c59/source/common/json/json_loader.cc#L520

I couldn't find related issues, tests in rapidjson support this construction. Im not sure I want to spend too much time on a rapidjson issue

```
external/com_github_tencent_rapidjson/include/rapidjson/schema.h:1546:39: error: Potential memory leak [clang-analyzer-unix.Malloc,-warnings-as-errors]
        CreateSchemaRecursive(&root_, PointerType(), document, document);
                                      ^
source/common/json/json_internal_legacy.cc:511:3: note: Taking false branch
  if (schema_document.Parse<0>(schema.c_str()).HasParseError()) {
  ^
source/common/json/json_internal_legacy.cc:517:29: note: Calling constructor for 'GenericSchemaDocument<rapidjson::GenericValue<rapidjson::UTF8<char>, rapidjson::MemoryPoolAllocator<rapidjson::CrtAllocator> >, rapidjson::CrtAllocator>'
  rapidjson::SchemaDocument schema_document_for_validator(schema_document);
                            ^
external/com_github_tencent_rapidjson/include/rapidjson/schema.h:1535:14: note: Field 'allocator_' is null
        if (!allocator_)
             ^
external/com_github_tencent_rapidjson/include/rapidjson/schema.h:1535:9: note: Taking true branch
        if (!allocator_)
        ^
external/com_github_tencent_rapidjson/include/rapidjson/schema.h:1539:24: note: 'uri' is null
        uri_.SetString(uri ? uri : noUri, uriLength, *allocator_);
                       ^
external/com_github_tencent_rapidjson/include/rapidjson/schema.h:1539:24: note: '?' condition is false
external/com_github_tencent_rapidjson/include/rapidjson/schema.h:1541:46: note: Calling 'CrtAllocator::Malloc'
        typeless_ = static_cast<SchemaType*>(allocator_->Malloc(sizeof(SchemaType)));
                                             ^
external/com_github_tencent_rapidjson/include/rapidjson/internal/../allocators.h:79:13: note: 'size' is 336
        if (size) //  behavior of malloc(0) is implementation defined.
            ^
external/com_github_tencent_rapidjson/include/rapidjson/internal/../allocators.h:79:9: note: Taking true branch
        if (size) //  behavior of malloc(0) is implementation defined.
        ^
external/com_github_tencent_rapidjson/include/rapidjson/internal/../allocators.h:80:20: note: Memory is allocated
            return std::malloc(size);
                   ^
external/com_github_tencent_rapidjson/include/rapidjson/schema.h:1541:46: note: Returned allocated memory
        typeless_ = static_cast<SchemaType*>(allocator_->Malloc(sizeof(SchemaType)));
                                             ^
external/com_github_tencent_rapidjson/include/rapidjson/schema.h:1546:39: note: Potential memory leak
        CreateSchemaRecursive(&root_, PointerType(), document, document);
                                      ^
```

**Comment by asraa** [comment]
Only failing existing clang-tidy. Would appreciate any comment! (especially around the new way of holding line_number_)

**Comment by asraa** [comment]
Would it be helpful if I displayed the diff between the old/new json loader?

The full diff is here https://gist.github.com/asraa/67c51ac7ab48570e471a744fb4c02068

The main differences are:
* renaming: SAX callbacks are named in nlohmann as `start_object` instead of `StartObject` and have slightly different signatures (all unused params though), `asRapidJsonDocument` is renamed to `asJsonDocument`.
* The new Json `ObjectHandler` holds on directly to an updating `line_number_` instead of holding on to a `LineCountingStringStream` with a line number member
* The new custom JSON "stream" that counts line numbers is implemented as a custom container that increments line numbers whenever the character at a position is a newline
* nlohmann/json had no API to push back items onto a json document with an allocator parameter
* No schema validation
* Slight differences in error messaging formatting. nlohmann/json will also give you the token at the error.

**Comment by asraa** [comment]
ping @jmarantz this is ready. 
clang-tidy will fail on an existing rapidJSON issue (see https://github.com/envoyproxy/envoy/pull/14467#issuecomment-758016926).

File: source/common/json/json_internal.cc [hunk 5]
```diff
@@ -0,0 +1,668 @@
+#include "common/json/json_internal.h"
```
Comment (jmarantz) [inline source/common/json/json_internal.cc:1]: qq: this file is all new code, and so needs to be reviewed line-by-line? This is not a copy of of something existing?

File: source/common/json/json_internal.h [hunk 6]
```diff
@@ -0,0 +1,22 @@
+#pragma once
+
+#include <list>
```
Comment (jmarantz) [inline source/common/json/json_internal.h:3]: `list` is not used in this header.

File: source/common/json/json_internal_legacy.h [hunk 7]
```diff
@@ -0,0 +1,22 @@
+#pragma once
+
+#include <list>
```
Comment (jmarantz) [inline source/common/json/json_internal_legacy.h:3]: `list` is not used in this file.

File: source/common/json/json_internal.cc (hunk 5 above)
Comment (asraa) [reply source/common/json/json_internal.cc:1]: Sort of, the diff is above (between the legacy and this one). They have some major differences (listed above) that made it difficult to inherit from the same SAX parser.

File: docs/root/version_history/current.rst [hunk 8]
```diff
@@ -42,11 +42,13 @@ Removed Config or Runtime
 
 New Features
 ------------
+
 * access log: added the :ref:`formatters <envoy_v3_api_field_config.core.v3.SubstitutionFormatString.formatters>` extension point for custom formatters (command operators).
 * access log: support command operator: %REQUEST_HEADERS_BYTES%, %RESPONSE_HEADERS_BYTES% and %RESPONSE_TRAILERS_BYTES%.
 * dispatcher: supports a stack of `Envoy::ScopeTrackedObject` instead of a single tracked object. This will allow Envoy to dump more debug information on crash.
 * http: added support for :ref:`:ref:`preconnecting <envoy_v3_api_msg_config.cluster.v3.Cluster.PreconnectPolicy>`. Preconnecting is off by default, but recommended for clusters serving latency-sensitive traffic, especially if using HTTP/1.1.
 * http: change frame flood and abuse checks to the upstream HTTP/2 codec to ON by default. It can be disabled by setting the `envoy.reloadable_features.upstream_http2_flood_checks` runtime key to false.
+* json: introduced new JSON parser (https://github.com/nlohmann/json) to replace RapidJSON. The new parser is enabled by default. To revert to the legacy RapidJSON parser, enable the runtime feature `envoy.reloadable_features.legacy_json`.
```
Comment (jmarantz) [inline docs/root/version_history/current.rst:51]: consider making the initial check-in leave the existing default, and then do another PR to flip the default. That way if a problem is discovered, we won't need to roll back the whole PR.

File: source/common/json/json_internal.cc [hunk 9]
```diff
@@ -0,0 +1,668 @@
...
+  bool binary(binary_t&) override { return false; }
+  bool parse_error(std::size_t, const std::string& token,
+                   const nlohmann::detail::exception& ex) override {
+    // Errors are formatted like "[json.exception.parse_error.101] parse error at line x, column y:
+    // error string."
+    absl::string_view error = ex.what();
+    // Extract position information (line, column, token) in the error.
+    auto start = error.find("line");
+    auto npos = error.find(":");
+    error_position_ = absl::StrCat(error.substr(start, npos - start), ", token ", token);
```
Comment (jmarantz) [inline source/common/json/json_internal.cc:207]: would like to see some checking of these `find()` results against `std::npos` or `absl::npos`. And while we're at it, let's not also have a local called `npos` :)

File: source/common/json/json_internal.cc [hunk 10]
```diff
@@ -0,0 +1,668 @@
...
+  return j;
+}
+
+uint64_t Field::hash() const { return HashUtil::xxHash64(asJsonString()); }
+
+bool Field::getBoolean(const std::string& name) const {
+  checkType(Type::Object);
+  auto value_itr = value_.object_value_.find(name);
+  if (value_itr == value_.object_value_.end() || !value_itr->second->isType(Type::Boolean)) {
+    throw Exception(fmt::format("key '{}' missing or not a boolean from lines {}-{}", name,
```
Comment (jmarantz) [inline source/common/json/json_internal.cc:362]: what's our general strategy to avoid this occurring during request processing?

File: source/common/json/json_internal.cc [hunk 11]
```diff
@@ -0,0 +1,668 @@
...
+  }
+  return value_itr->second->doubleValue();
+}
+
+double Field::getDouble(const std::string& name, double default_value) const {
+  checkType(Type::Object);
+  auto value_itr = value_.object_value_.find(name);
+  if (value_itr != value_.object_value_.end()) {
+    return getDouble(name);
+  } else {
```
Comment (jmarantz) [inline source/common/json/json_internal.cc:393]: I think we prefer the form
```
if (exceptional-condition) {
  return exceptional-value;
}
return normal value;
```
But I can't find a style-guide ref for that. THere's a few different instances of that in this file. `Field::getDouble()` above is an example where you *are* doing it the way I think we prefer.

**Review by jmarantz — COMMENTED** [review]
Feel free to ignore the style nits if this is not really code you edited.

Basically this looks good. Will need main merge.

File: source/common/json/json_internal.cc (hunk 10 above)
Comment (asraa) [reply source/common/json/json_internal.cc:362]: There is one usage of `loadFromString` in the core dataplane code here, as noted in the description I will swap this with protobuf's json parser in a future PR
> There is one usage in core to parse metadata in the HeaderFormatter:

These get[String/Boolean] methods that can throw are largely unused, but there are <10 usages in filters that I will replace with those with defaults that don't throw in that PR.

File: source/common/json/json_internal.cc (hunk 11 above)
Comment (asraa) [reply source/common/json/json_internal.cc:393]: existing style, fixed in the legacy code as well.

File: source/common/json/json_internal.cc (hunk 9 above)
Comment (asraa) [reply source/common/json/json_internal.cc:207]: done, i added a reference, there should always be a ": error string" but I found in the documentation that position information may not be present, so that's handled now.

Thank you!

File: docs/root/version_history/current.rst (hunk 8 above)
Comment (asraa) [reply docs/root/version_history/current.rst:51]: like that idea very much. done

**Review by jmarantz — APPROVED** [review]

**Comment by asraa** [comment]
@envoyproxy/senior-maintainers  Could you PTAL? This adds the code for a new json parser, does not flip the switch.

**Review by mattklein123 — APPROVED** [review]
LGTM at a high level. Very excited to see this moving forward!
